### ISSUE TITLE
Cookbook/verified examples for gpt-oss

### ISSUE THREAD

**Issue by jphme** [thinking-machines-lab/tinker-feedback#63]
I now burned a few hundred dollars (and >3bn tokens) on your API and still didn't manage to get  gpt-oss to learn properly.

We have a case thats quite similar to the `search_tool` example with multi-turn search and interleaved reasoning, however longer rollouts (up to 30k tokens) than in the example. 

There are several issues:
* the reference renderer from the cookbook has some deviations from OpenAIs reference harmony cookbook (e.g. different system prompt) - not sure though whether this is a major problem
* The KL is extremely high and fluctuates wildly, despite proper masking and token accuracy (see also #22 ) 
* It´s still unclear for me, what the optimal learning rate is (I got ~some~ learning with 1e-4)

Any plans to release a cookbook for gpt-oss that properly uses the harmony renderer over multiple turns and with verified results?

Thanks 👍

**Comment by opherlieber** [comment]
Hi,

- Can you share what deviation you see from harmony system prompt and which cookbook renderer you are using? The 'gpt_oss_medium_reasoning' renderer should match the default official system-prompt (There is a test in the cookbok that validates it against the HF chat template, unless that one is also wrong)
- How high/fluctuating is the KL? Could you share the values/chart?
- 1e-4 might be a bit high, we have some succesful RL runs (e.g. on deepmath-rl) with 4e-5, so I would try that and maybe 2e-5 if you haven't already. Will also try 1e-4 on my side to see how it behaves and let you know.
- We are looking into improving the cookbook with more examples

**Comment by jphme** [comment]
> * Can you share what deviation you see from harmony system prompt and which cookbook renderer you are using? The 'gpt_oss_medium_reasoning' renderer should match the default official system-prompt (There is a test in the cookbok that validates it against the HF chat template, unless that one is also wrong)

We used `renderers/gpt-oss.py` - is there a different one?
 
The main difference was in the system prompt with tools. If i understand correctly, this is not tested against HF (see [here](https://github.com/thinking-machines-lab/tinker-cookbook/blob/9934ede014f1f20f76460ab9ca9e6014540ab2f0/tinker_cookbook/tests/test_renderers.py#L397) ) and maybe we were just confused by [this comment](https://github.com/thinking-machines-lab/tinker-cookbook/blob/9934ede014f1f20f76460ab9ca9e6014540ab2f0/tinker_cookbook/renderers/gpt_oss.py#L627) but the rendered system message (in default config) only included 

`"Calls to these tools must go to the commentary channel: 'functions'."`

instead of the full one (starting with `"You are ChatGPT, a large language model...` ). 

However as we iterated a lot on the code and renderer would have to create a new example for this - might this be indeed an error? (can try to replicate + fix - are you accepting prs?)

> * How high/fluctuating is the KL? Could you share the values/chart?

Here from a recent run, definitely doesnt look right ;-) (could be different issues though) - also it fluctuates a lot between different batches:

<img width="958" height="309" alt="Image" src="https://github.com/user-attachments/assets/b3945886-0369-439b-9f33-0eaf26f939a3" />

> * 1e-4 might be a bit high, we have some succesful RL runs (e.g. on deepmath-rl) with 4e-5, so I would try that and maybe 2e-5 if you haven't already. Will also try 1e-4 on my side to see how it behaves and let you know.
> * We are looking into improving the cookbook with more examples

Thanks a lot, looking forward to it. We want to preserve the harmony format instead simpler think/tool tags as in the qwen examples; this (or some error connected to it) is probably the root of most issues and where a working example (especially for a tool-calling RL) would help a lot 👍 .

**Comment by opherlieber** [comment]
Yes, `renderers/gpt-oss.py` is the right one. `gpt_oss_medium_reasoning` is the option in `get_renderer` that instantiates `GptOssRenderer` with medium reasoning effort and enables the system prompt. I think you probably do want to enable system-prompt even with that tool message, so need to check where that comment came from. We should probably add a test that checks gpt-oss tool-calling rendering vs harmony. 
We do accept PR's to tinker-cookbook if you have any additions/fixes !

That KL seems high, also at step0, so likely not related to LR. Possibly you have some very high logprobs in there which can increase KL a lot even if it's the same relative difference (Maybe due to rendering issues). If you can share sampler and trainer logprobs for that step0 that would be great (Assuming you can't share the actual training data. But if you can share some more general training data that reproduces this KL it can also be helpful)

**Comment by jphme** [comment]
> That KL seems high, also at step0, so likely not related to LR. Possibly you have some very high logprobs in there which can increase KL a lot even if it's the same relative difference (Maybe due to rendering issues). If you can share sampler and trainer logprobs for that step0 that would be great (Assuming you can't share the actual training data. But if you can share some more general training data that reproduces this KL it can also be helpful)

@opherlieber  thanks! Can't share the raw data but here is part of the analysis claude did for me:
https://gist.github.com/jphme/436b82acc0de26ce43cabe0f97e7e898 (see also the [second, newer part](https://gist.github.com/jphme/436b82acc0de26ce43cabe0f97e7e898#kl-token-investigation-report-2026-01-13)). 

And here is a [log/debug output](https://gist.github.com/jphme/668c2ce4445ef7a34b16e1b40b19ada4) from step 1.

Basically for single-turn/no tool-calling everything looks well and KL low as expected but for harmony-formatted multi-turn/toolcalling something seems to be broken (also tested multiple renderers) - and I wasn't able to figure out what, despite multiple long debugging sessions and claude also working away on it. Learning rate doesnt change anything.

Many thanks for looking into it!

**Comment by opherlieber** [comment]
Hi,
Trying to repro this. 
How are you calculating train/kl_v2 in your chart?
In the log you shared it looks like you are getting average absdiff's of around ~0.1-0.24, which would translate to kl_v2 of 0.005-0.028 (Based on how we calculate it in the cookbook, i.e. mse(diffs)*0.5), which still isn't ideal but not that far off the expected range.
